core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolver.java

```diff
@@ -0,0 +1,306 @@
+                return;
+            }
+
+            @SuppressWarnings("unchecked")
+            final List<DnsRecord> records = (List<DnsRecord>) f.getNow();
+            InetAddress inetAddress = null;
+            long ttlMillis = -1;
+            try {
+                for (DnsRecord r : records) {
+                    final byte[] addrBytes = extractAddressBytes(r, logger, hostname);
+                    if (addrBytes == null) {
+                        continue;
+                    }
+                    try {
+                        inetAddress = InetAddress.getByAddress(hostname, addrBytes);
+                        ttlMillis = TimeUnit.SECONDS.toMillis(
+                                Math.max(Math.min(r.timeToLive(), maxTtl), minTtl));
+                        break;
+                    } catch (UnknownHostException e) {
+                        // Should never reach here because we already validated it in extractAddressBytes.
+                        result.completeExceptionally(
+                                new IllegalArgumentException("Invalid address: " + hostname, e));
+                        return;
+                    }
+                }
+            } finally {
+                records.forEach(ReferenceCountUtil::safeRelease);
+            }
+
+            if (inetAddress == null) {
+                result.completeExceptionally(
+                        new UnknownHostException("failed to receive DNS records for " + hostname));
+                return;
+            }
+            final CacheEntry entry = new CacheEntry(inetAddress, ttlMillis, questions);
+            result.complete(entry);
+            entry.scheduleRefresh(ttlMillis);
+        });
+    }
+
+    @Override
+    protected void doResolveAll(InetSocketAddress unresolvedAddress, Promise<List<InetSocketAddress>> promise)
+            throws Exception {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Closes all the resources allocated and used by this resolver.
+     *
+     * <p>Please note that this method does not clear the {@link CacheEntry} because the {@link #cache} is
+     * created by {@link RefreshingAddressResolverGroup} and shared across from all
+     * {@link RefreshingAddressResolver}s. {@link CacheEntry} is cleared when
+     * {@link RefreshingAddressResolverGroup#close()} is called.
+     */
+    @Override
+    public void close() {
+        resolverClosed = true;
+        resolver.close();
+    }
+
+    final class CacheEntry implements Runnable {
+
+        private final InetAddress address;
+        private final long ttlMillis;
+        private final List<DnsQuestion> questions;
+
+        /**
+         * No need to be volatile because updated only by the {@link #executor()}.
+         */
+        private int numAttemptsSoFar = 1;
+
+        private volatile boolean servedFromCache;
+
+        @VisibleForTesting
+        @Nullable
+        ScheduledFuture<?> refreshingScheduledFuture;
```

Comment:
I think I'd name it `refreshFuture` - `scheduled` is already part of the type and this future only runs once until rescheduling instead of using `scheduleWithFixedDelay` for example

Reply:
Fixed. ;)